server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreSnapshotMetrics.java

```diff
@@ -0,0 +1,93 @@
+
+    private final SnapshotMetrics snapshotMetrics;
+    private final CounterMetric shardSnapshotsInProgress = new CounterMetric();
+    private final CounterMetric snapshotRateLimitingTimeInNanos = new CounterMetric();
+    private final CounterMetric restoreRateLimitingTimeInNanos = new CounterMetric();
+    private final CounterMetric numberOfBlobsUploaded = new CounterMetric();
+    private final CounterMetric numberOfBytesUploaded = new CounterMetric();
+    private final CounterMetric uploadTimeInNanos = new CounterMetric();
+    private final CounterMetric uploadReadTimeInNanos = new CounterMetric();
+    private final CounterMetric numberOfShardSnapshotsStarted = new CounterMetric();
+    private final CounterMetric numberOfShardSnapshotsCompleted = new CounterMetric();
+    private final Map<String, Object> metricAttributes;
+
+    public BlobStoreSnapshotMetrics(ProjectId projectId, RepositoryMetadata repositoryMetadata, SnapshotMetrics snapshotMetrics) {
+        this.snapshotMetrics = snapshotMetrics;
+        metricAttributes = SnapshotMetrics.createAttributesMap(projectId, repositoryMetadata);
+    }
+
+    public void incrementSnapshotRateLimitingTimeInNanos(long throttleTimeNanos) {
+        snapshotMetrics.snapshotCreateThrottleDurationCounter().incrementBy(throttleTimeNanos);
+        snapshotRateLimitingTimeInNanos.inc(throttleTimeNanos);
+    }
+
+    public long snapshotRateLimitingTimeInNanos() {
+        return snapshotRateLimitingTimeInNanos.count();
+    }
+
+    public void incrementRestoreRateLimitingTimeInNanos(long throttleTimeNanos) {
+        snapshotMetrics.snapshotRestoreThrottleDurationCounter().incrementBy(throttleTimeNanos);
+        restoreRateLimitingTimeInNanos.inc(throttleTimeNanos);
+    }
+
+    public long restoreRateLimitingTimeInNanos() {
+        return restoreRateLimitingTimeInNanos.count();
+    }
+
+    public void incrementCountersForPartUpload(long partSizeInBytes, long partWriteTimeNanos) {
+        snapshotMetrics.snapshotBytesUploadedCounter().incrementBy(partSizeInBytes);
+        snapshotMetrics.snapshotUploadDurationCounter().incrementBy(partWriteTimeNanos);
+        numberOfBytesUploaded.inc(partSizeInBytes);
+        uploadTimeInNanos.inc(partWriteTimeNanos);
+    }
+
+    public void incrementNumberOfBlobsUploaded() {
+        snapshotMetrics.snapshotBlobsUploadedCounter().increment();
+        numberOfBlobsUploaded.inc();
+    }
+
+    public void shardSnapshotStarted() {
+        snapshotMetrics.snapshotsShardsStartedCounter().increment();
+        numberOfShardSnapshotsStarted.inc();
+        shardSnapshotsInProgress.inc();
+    }
+
+    public void shardSnapshotCompleted(IndexShardSnapshotStatus status) {
+        snapshotMetrics.snapshotsShardsCompletedCounter().increment();
+        if (status.getStage() == IndexShardSnapshotStatus.Stage.DONE) {
+            snapshotMetrics.snapshotShardsDurationHistogram().record(status.getTotalTime() / 1_000f);
+        }
```

Comment:
totalTime is only set when the shard is complete, failed or cancelled. I think we only want to count completed ones?

Reply:
Yes, good idea, fixed that and some other attribute issues here 0a19687092fa494f6838b6567f48731fa9e4c69e